django/core/mail/message.py

```diff
@@ -362,68 +217,69 @@ def attach_file(self, path, mimetype=None):
-        return self._create_attachments(msg)
+    def _add_bodies(self, msg):
+        if self.body or not self.attachments:
+            encoding = self.encoding or settings.DEFAULT_CHARSET
+            body = force_str(
+                self.body or "", encoding=encoding, errors="surrogateescape"
+            )
+            msg.set_content(body, subtype=self.content_subtype, charset=encoding)
 
-    def _create_attachments(self, msg):
+    def _add_attachments(self, msg):
         if self.attachments:
-            encoding = self.encoding or settings.DEFAULT_CHARSET
-            body_msg = msg
-            msg = SafeMIMEMultipart(_subtype=self.mixed_subtype, encoding=encoding)
-            if self.body or body_msg.is_multipart():
-                msg.attach(body_msg)
+            if hasattr(self, "mixed_subtype"):
+                # RemovedInDjango70Warning
+                raise AttributeError(
+                    "EmailMessage no longer supports the"
+                    " undocumented `mixed_subtype` attribute"
+                )
+            msg.make_mixed()
             for attachment in self.attachments:
-                if isinstance(attachment, MIMEBase):
+                if isinstance(attachment, email.message.MIMEPart):
+                    msg.attach(attachment)
+                elif isinstance(attachment, MIMEBase):
+                    # RemovedInDjango70Warning
                     msg.attach(attachment)
                 else:
-                    msg.attach(self._create_attachment(*attachment))
-        return msg
-
-    def _create_mime_attachment(self, content, mimetype):
-        """
-        Convert the content, mimetype pair into a MIME attachment object.
+                    self._add_attachment(msg, *attachment)
 
-        If the mimetype is message/rfc822, content may be an
-        email.Message or EmailMessage object, as well as a str.
-        """
-        basetype, subtype = mimetype.split("/", 1)
-        if basetype == "text":
-            encoding = self.encoding or settings.DEFAULT_CHARSET
-            attachment = SafeMIMEText(content, subtype, encoding)
-        elif basetype == "message" and subtype == "rfc822":
-            # Bug #18967: Per RFC 2046 Section 5.2.1, message/rfc822
-            # attachments must not be base64 encoded.
+    def _add_attachment(self, msg, filename, content, mimetype):
+        encoding = self.encoding or settings.DEFAULT_CHARSET
+        if mimetype is None:
+            mimetype = DEFAULT_ATTACHMENT_MIME_TYPE
+        maintype, subtype = mimetype.split("/", 1)
+        # See email.contentmanager.set_content() docs for the cases here.
+        if maintype == "text":
+            # For text/*, content must be str, and maintype cannot be provided.
+            if isinstance(content, bytes):
+                content = content.decode()
+            msg.add_attachment(
+                content, subtype=subtype, filename=filename, charset=encoding
+            )
```

Comment:
```suggestion
        maintype, subtype = mimetype.split("/", 1)
        # See email.contentmanager.set_content() docs for the cases here.
        if maintype == "text":
            msg.add_attachment(
                content, subtype=subtype, filename=filename, charset=encoding
            )
```
I can update this to the above without test failures

Reply:
So I've gone back and forth on this a couple of times. There's definitely a missing test case and a slight bug in my implementation.

I've concluded that there's also an existing bug where setting EmailMessage.attachments doesn't work the same as EmailMessage.attach() for utf-8 encoded text/* content.

I've prepended a commit that fixes that (and adds test cases). With that, the diff in this part of the code is clearer, and your suggested update would break the new tests.

Reply:
Done: ticket-36478.

If that's not accepted, the alternative would be to check for this case here and raise a useful error. (Passing bytes with a text/* mimetype to EmailMessage.add_attachment causes a cryptic error.)